### PR TITLE
[2.0.x] Use segmented moves when moving axis from LCD

### PR DESCRIPTION
**Pull request by tcm0116**
Currently, move commands generated by the LCD Move axis sub-menu are sent directly to the `Planner` class as unsegmented moves on Delta machines. This can be a problem as it can cause the nozzle to drag into the bed when commanding large moves in X or Y while the nozzle is close to the bed.

This change is a bit complicated because of the way the LCD previously modified the `current_position` array in order to generate the movement command. `prepare_move_to_destination` compares `current_position` with `destination` to determine if any movements are commanded in the XY plane, and only segments the move if so. By setting `current_position` to the actual desired position, it breaks the ability to determine if a movement is being commanded in the XY plane. This PR fixes once such instance of this erroneous  method of commanding movements, but there are several other places where it is done. I'm working on another PR that cleans all of that up and makes all movement commands that could be in the XY plane use segmented motion.

This PR should be easily adaptable to 1.1.x if so desired.

### CODE REVIEW

**Comment by thinkyhead** [comment]
I'll start to wrap up that 1.1.x version. Otherwise, this looks awesome!

**Comment by thinkyhead** [comment]
@tcm0116 Theoretically, the motion behavior from the LCD could be done as a cumulative function, where all axes are allowed to be moved freely, except the `manage_manual_move` function would just only do one axis at a time in some priority order. For example, if it saw there was motion in all axes —X, Y, Z, and E— it would run out all E motion first, then all Z motion, and then, I suppose, combine all XY motion. This would get rid of the need to have the `manual_move_axis` (axis that was just moved) variable.

**Comment by tcm0116** [comment]
True, but the user would be hard pressed to switch the desired axis and command a movement before the queued segments were completed. I tried :-)​

**Comment by thinkyhead** [comment]
> True, but the user would be hard pressed to switch the desired axis and command a movement before the queued segments were completed. I tried :-)​

I kind of think it's irrelevant. The natural interaction with the LCD will tend to be with a single axis until movement is done. If the axis is switched, then we should expect previous moves to finish. But, if an E axis move were added, I guess it would supersede any leftovers on other axes… In all things, the question I have is "how jarring would it be?" I think little quirks like these would be tolerable, if consistent and sensible. So, we have the option to simplify at that level. Or, just keep going with single-axis as this does, which is also fine. Neither is all that expensive.

**Comment by tcm0116** [comment]
Are you referring to `processing_manual_move` when you say "axis just moved" variable?

**Comment by thinkyhead** [comment]
The `manual_move_axis` variable indicates the axis that was moved.

**Comment by thinkyhead** [comment]
The `manual_move_axis` variable indicates the axis that was moved. Since `manual_move_axis == NO_AXIS` is the same as `!processing_manual_move`, you probably don't need to add that new flag.

**Comment by tcm0116** [comment]
An alternative implementation could be to have `prepare_kinematic_move_to` return **true** if there's not enough space in the planner's buffer to store the segments, and then have `prepare_move_to_destination` return the status. This would prevent `Planner::_buffer_line` from calling `idle()`. We revert `manage_manual_move` to the current implementation except for adding a check for the return value of `prepare_move_to_destination`.

**Comment by thinkyhead** [comment]
@tcm0116 I've patched this up a bunch and simplified the implementation a little, based on results from testing with 1.1.x and comparing the difference in cartesian vs kinematic.

The only side-effect that gets coordinates out of sync is the fact that `prepare_move_to_destination` always calls `set_current_to_destination` when it's done.

So, it's true that you can't send new moves to `prepare_move_to_destination` recursively — that gets ugly fast. However, as long as the `current_position` value is updated _**just before**_ calling `prepare_move_to_destination`, the LCD loop can continue to accumulate moves and it will show the right value. The move will just wait until space becomes available in the planner, as usual.

See if the updated code works for you without any unwanted side-effects.

**Comment by tcm0116** [comment]
@thinkyhead see my comments in #7917 regarding setting `current_position` prior to calling `set_destination_to_current`. That method will result in an single unsegmented call to `Planner::buffer_line` because of the following lines in `prepare_kinematic_move_to `:

https://github.com/MarlinFirmware/Marlin/blob/5ad8f5c306c1760d5dc62b58bac962c62b776b1f/Marlin/src/module/motion.cpp#L517-L521

**Comment by thinkyhead** [comment]
Patched up further, so it's pretty much back where we started! But, only applies the changes for kinematic and leaves the old code in place for Cartesian.

File: Marlin/src/lcd/ultralcd.cpp [hunk 1]
```diff
@@ -2800,17 +2840,33 @@ void kill_screen(const char* lcd_msg) {
...
-      lcd_implementation_drawedit(name, move_menu_scale >= 0.1 ? ftostr41sign(current_position[axis]) : ftostr43sign(current_position[axis]));
+    encoderPosition = 0;
+    if (lcdDrawUpdate && !processing_manual_move) {
+      const float pos = current_position[axis]
+        #if IS_KINEMATIC
+          + manual_move_offset
+        #endif
+      ;
+      lcd_implementation_drawedit(name, move_menu_scale >= 0.1 ? ftostr41sign(pos) : ftostr43sign(pos));
+    }
```
Comment (thinkyhead) [inline Marlin/src/lcd/ultralcd.cpp]: @tcm0116 — I'm trying to reconcile code that differs between the branches. Do you think this code is more correct than the [equivalent code in the 1.1.x branch](https://github.com/MarlinFirmware/Marlin/pull/7917/files#diff-34f7cdd618621ae7d8c113cf0b86a2f4R2930)?

File: Marlin/src/lcd/ultralcd.cpp [hunk 2]
```diff
@@ -2800,17 +2840,33 @@ void kill_screen(const char* lcd_msg) {
...
 
       manual_move_to_current(axis);
 
-      encoderPosition = 0;
       lcdDrawUpdate = LCDVIEW_REDRAW_NOW;
     }
-    if (lcdDrawUpdate)
-      lcd_implementation_drawedit(name, move_menu_scale >= 0.1 ? ftostr41sign(current_position[axis]) : ftostr43sign(current_position[axis]));
+    encoderPosition = 0;
+    if (lcdDrawUpdate && !processing_manual_move) {
```
Comment (thinkyhead) [inline Marlin/src/lcd/ultralcd.cpp:2862]: @tcm0116 — I'm trying to reconcile code that differs between the branches. Do you think this code is more correct than the [equivalent code in the 1.1.x branch](https://github.com/MarlinFirmware/Marlin/pull/7917/files#diff-34f7cdd618621ae7d8c113cf0b86a2f4R2875)?

**Pull request merged**